### DISCUSSION TITLE
시간복잡도

### DISCUSSION THREAD

**Discussion by jangwooooo** [subak-ba/cs-study#21] [자료 구조]
# 시간복잡도 (time complexity)

## 정의
시간복잡도란 어떠한 알고리즘의 성능을 설명하는 값이다.

<br>

## 측정법
실행시간이 아닌 반복횟수를 측정한다.
그 이유는 실행시간은 컴퓨터 사양이나 프로그래밍 언어에 따라 차이가 있기 때문이다.

<br>

## 표기법
입력값의 크기에 따른 함수의 증가량, 이것을 성장률이라고 부른다.
이때 중요하지 않는 상수와 계수들을 제거하면 알고리즘의 실행시간에서 중요한 성장률에 집중할 수있다.
가장 큰 영향을 주는 항만 계산하는 것을 점금적 표기법이라 부른다.

점근적 표기법은 3가지가 있다.
최상의 경우 : 오메가 표기법 (Big-Ω Notation)
평균의 경우 : 세타 표기법 (Big-θ Notation)
최악의 경우 : 빅오 표기법 (Big-O Notation)

평균인 세타 표기가 가장 정확하지만 측정하기가 어렵다.
그래서 보통 최악의 경우인 빅오 표기법을 사용한다.

```
n + 4n^2        O(n^2)   --> n^2 이 가장 큰영향을 미친다.
2n + 15         O(n) --> n이 가장 큰영향을 미친다.
```
입력값의 크기에 따라 함수의 증가량이 큰 항만을 계산하고 빅오 표기법으로 나타낸 것이다.

### Big-O 차트
![image](https://github.com/subak-ba/cs-study/assets/108380536/891b7912-fdbb-4e42-8086-f8144ce3825b)

### 정렬 알고리즘 비교
![image](https://github.com/subak-ba/cs-study/assets/108380536/7c878544-2f00-4c06-9b18-78264eabc1fd)

<br/>

## 예제
```java
for(int i = 0; i < n; i++) {
    sum += i;
}

for (int i = 0; i < 4; I++) {
    for (int j = 0; j < n; j++) {
        for (int k = 0; k < n; k++) {
            sum += k;
        }
    }
}
```
위 코드는 n + 4(n x n) 반복되지만 가장 큰 영향을 미치는 항만을 계산하므로 O(n^2)의 시간복잡도를 가진다.

```java
for(int i = 0; i < 2; i++) {
    for (int j = 0; j < n; j++) {
        sum += j;
    }
}

for(int i = 0; i < 15; i++) {
    sum += i;
}
```
위 코드는 2n + 15번 반복되지만 가장 큰 영향을 미치는 항만을 계산하므로 O(n)의 시간복잡도를 가진다.

**Comment by onee-only** [comment]
👍

**Comment by vumrra** [comment]
<img width="673" alt="스크린샷 2024-03-30 오후 6 29 29" src="https://github.com/subak-ba/cs-study/assets/131235625/c92152c2-e22e-4549-8141-b53c5fc3f53e">

복잡도라는 큰 주제에서 시간복잡도, 공간복잡도 나뉘는거 같아요. 본 주제와 관련이 없을 수 있지만 공간 복잡도에 대한 내용도 추가되면 좋을거같아요!

**Comment by BLoHny** [comment]
다른 표현식도 예시가 있었다면 더 좋앆을 것 같습니다!

👍
